src/merkle.rs

```diff
@@ -237,19 +549,109 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>, U: Unsigned> MerkleTree<T, A, K,
         );
 
         Ok(MerkleTree {
-            data,
+            data: Some(data),
             leafs,
+            len: tree_len,
             height,
+            // FIXME: correct?
+            //top_layer_nodes: N::to_usize(),
+            //sub_tree_layer_nodes: R::to_usize(),
+            top_layer_nodes: 0,
+            trees: None,
+            sub_tree_layer_nodes: 0,
+            sub_trees: None,
             root,
             _u: PhantomData,
             _a: PhantomData,
             _t: PhantomData,
+            _n: PhantomData,
+            _r: PhantomData,
         })
     }
 
+    /// Generate merkle sub tree inclusion proof for leaf `i` for
+    /// either the top layer or the sub-tree layer, specified by the
+    /// top_layer flag
+    #[inline]
+    pub fn gen_sub_tree_proof(
+        &self,
+        i: usize,
+        top_layer: bool,
```

Comment:
I would prefer two different methods, instead of a boolean flag, at least for the public api. Alternatively at least an enum instead of a bool, to make it more descriptive

Reply:
Actually, this should not be public.

Reply:
Same with `gen_cached_subtree_proof`. I'll just remove pub on them both, unless you also strongly prefer an alternative.

Reply:
that's okay if these are not public